### PR TITLE
[12.0][IMP] shopinvader_cart_expiry: Add expiration unit

### PR DESCRIPTION
**Pull request by rousseldenis**
Improvement to take into account some particular usages of
cart expiration delays.

The service that retrieves cart, includes now the expiration date.

### CODE REVIEW

File: shopinvader_cart_expiry/models/sale_order.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    @api.depends(
+        "shopinvader_backend_id.cart_expiry_delay", "typology", "write_date"
+    )
+    def _compute_cart_expiration_date(self):
+        """
+        Compute the cart expiration date
+        :return:
+        """
+        carts = self.filtered(
```
Comment (simahawk) [inline shopinvader_cart_expiry/models/sale_order.py]: this way you loop twice on all of them.
I'd do:
```py
def _is_expiring_cart(self):
    return self.shopinvader_backend_id and self.shopinvader_backend_id.cart_expiry_delay != 0 and self.typology == "cart"

for order in self:
    if order._is_expiring_cart():
[...]
```

File: shopinvader_cart_expiry/models/shopinvader_backend.py [hunk 2]
```diff
@@ -17,6 +17,11 @@ class ShopinvaderBackend(models.Model):
         "Let 0 (or less) to disable the feature.",
         default=0,
     )
+    cart_expiry_delay_unit = fields.Selection(
+        [("minutes", "Minute()s"), ("hours", "Hour(s)"), ("days", "Day(s)")],
```
Comment (simahawk) [inline shopinvader_cart_expiry/models/shopinvader_backend.py]:
```suggestion
        [("minutes", "Minute(s)"), ("hours", "Hour(s)"), ("days", "Day(s)")],
```

File: shopinvader_cart_expiry/tests/test_cart_expiry.py [hunk 3]
```diff
@@ -32,6 +39,28 @@ def test_cart_expiry_scheduler(self):
...
+            {"cart_expiry_delay": 1, "cart_expiry_policy": "cancel"}
+        )
+        now_method = "odoo.fields.Datetime.now"
+        so_date = fields.Datetime.from_string(self.sale.write_date)
+        today = fields.Datetime.to_string(so_date + timedelta(hours=5))
+        self.backend.write(
+            {"cart_expiry_delay": 1, "cart_expiry_policy": "cancel"}
+        )
+        now_method = "odoo.fields.Datetime.now"
+        with mock.patch(now_method) as mock_now:
```
Comment (simahawk) [inline shopinvader_cart_expiry/tests/test_cart_expiry.py:49]: would be better to use freezegun

**Review by simahawk — CHANGES_REQUESTED** [review]

File: shopinvader_cart_expiry/tests/test_cart_expiry.py (hunk 3 above)
Comment (rousseldenis) [reply shopinvader_cart_expiry/tests/test_cart_expiry.py:49]: Needs to add a dependency just for a test ?

Comment (simahawk) [reply shopinvader_cart_expiry/tests/test_cart_expiry.py:49]: well, the behavior w/ freezegun is not the same. You grant that whatever piece of code is involved is going to use the same date. Also, next time someone will add a test that requires a date patch, likely he/she's going to copy/paste this.
Anyway, not a blocker but TODO comment wouldn't hurt ;)

**Comment by rousseldenis** [comment]
@simahawk

**Review by simahawk — APPROVED** [review]

File: shopinvader_cart_expiry/tests/test_cart_expiry.py (hunk 3 above)
Comment (lmignon) [reply shopinvader_cart_expiry/tests/test_cart_expiry.py:49]: I also like freezgun... TODO would be nice.

**Review by lmignon — APPROVED** [review]
LGTM (code review)

File: shopinvader_cart_expiry/tests/test_cart_expiry.py (hunk 3 above)
Comment (simahawk) [reply shopinvader_cart_expiry/tests/test_cart_expiry.py:49]: another one https://github.com/shopinvader/odoo-shopinvader/pull/535#pullrequestreview-331231906 :)

**Pull request merged**